### PR TITLE
Recurrent and batched TRPO

### PR DESCRIPTION
**Pull request by muupan**
~Merge #431 before this PR.~

- Precompute action distributions for importance weighting and KL divergence as #430 did
- Support recurrent models as #431 did
- Support batch training

### CODE REVIEW

File: chainerrl/agents/trpo.py [hunk 1]
```diff
@@ -350,27 +418,94 @@ def _update_policy(self, dataset):
...
+            policy_rs = self.policy.concatenate_recurrent_states(
+                [ep[0]['recurrent_state'][0] for ep in dataset])
+
+        flat_distribs, _ = self.policy.n_step_forward(
+            seqs_states, recurrent_state=policy_rs, output_mode='concat')
+
+        if self.standardize_advantages:
+            mean_advs = xp.mean(flat_advs)
+            std_advs = xp.std(flat_advs)
+            flat_advs = (flat_advs - mean_advs) / (std_advs + 1e-8)
```
Comment (toslunar) [inline chainerrl/agents/trpo.py]: Non-recurrent code standardizes advantages before computing action distributions.  Could you make their order consistent?

**Comment by muupan** [comment]
Let me investigate the failure in pfn-public-ci/chainerrl.py3.gpu .

**Comment by muupan** [comment]
I added changes to address flexCI failurue:
- move `self.model` to gpu if policy is on gpu
- skip test cases with recurrent=True when either
  - gpu (because NStepLSTM with gpu does not support double backprop), or
  - chainer==7.0.0b3 (https://github.com/chainer/chainer/pull/8037)

Remaining flexCI failures are only flakey tests and #533

**Comment by muupan** [comment]
@toslunar Can you review the changes I made?

**Pull request merged**